### PR TITLE
UnknownHighwayTagCheck completed

### LINKED ISSUE #464: UnknownHighwayTagCheck
### Description
This check attempt to address unknown highway tags.

### Requirements
This section should specify the requirements for an Atlas object (or OSM feature) to be flagged by this check. This section is mainly for devs to get a sense of feature candidates for the source code.
1. Object must be a way or node
2. Object highway tag must reside within the HighwayTag.java enum.

#### True Positives 
[Node:313559095](https://www.openstreetmap.org/node/313559095) has a highway tag that is unknown (“highway=priority”) 

#### False Positives 
[Way:28274387](https://www.openstreetmap.org/way/28274387) has a known highway tag (“highway=trunk”) 

**Supported regions**
All countries

### PR DESCRIPTION
**Pull request by reichg**
### Description:

This check attempts to flag all highway tags that are unknown to the [osm wiki page](https://wiki.openstreetmap.org/wiki/Key:highway).

### Test Results:

| ISO | Total Flags | Sampled  | Sampling % | TP | FP | False Positive Rate |
|-----|-------------|----------|------------|----|----| ------------------- |
|  ARG   |    6         |    6      |     100%       |   6 |   0 | 0%   |
|  AUS   |    78         |    78      |     100%       |   78 |   0 | 0%   |
|  CHL   |    13         |    13      |     100%       |   13 |   0 | 0%   |
|  FIN   |    15         |    15      |     100%       |   15 |   0 | 0%   |
|  NOR   |    11         |    11      |     100%       |   11 |   0 | 0%   |
|  PER   |    24         |    24      |     100%       |   24 |   0 | 0%   |
|  SWE   |    127         |    127      |     100%       |   127 |   0 | 0%   |

### Notes
PR has been put in to the Atlas repository to add 3 highway tags that do not already exist which will lower the total flags when approved but the fpr will not change.

resolves #464

### CODE REVIEW

**Comment by vladlemberg** [comment]
@reichg, this is great! quick question. do we need to flag Nodes with known highway type but not applicable to Node ?
For example: Node with highway=residential. There are currently (global) 4013 Nodes with incorrect highway types.  
Analysis by navigable_highway_type  
image

Random Examples with OSM Nodes (id). 
7553108892
2520631572
5616532077
1210008466
6163227662
6163227668
7757156741
7757156742
5260791174
8236303276
6919453737
6349934559

**Comment by reichg** [comment]
If I understand correctly you're saying we might need to determine which highway tags are relative to Nodes or Ways respectively? Example: 

"bus_stop" would be Node related but not Way related.
"trunk" would be Way related and not Node related? 

Does this distinction matter?

**Comment by Bentleysb** [comment]
@reichg, I am guessing this PR should wait on https://github.com/osmlab/atlas/pull/723?

**Comment by reichg** [comment]
> @reichg, I am guessing this PR should wait on https://github.com/osmlab/atlas/pull/723?

Yeah I'll put this back in draft. I was working on similar logic to what I had in the Atlas repo but now I can drop that and use the Atlas additions when I complete that.

**Comment by ameliaewatts** [comment]
@reichg Because this is still using the previous Travis framework, in order to pass, we need to close it and reopen this pr or push a new commit!

**Comment by reichg** [comment]
> @reichg Because this is still using the previous Travis framework, in order to pass, we need to close it and reopen this pr or push a new commit!

Yeah I am just waiting on the atlas-checks' next release because this check is dependent on a PR that just got merged into the Atlas repo. [PR 723](https://github.com/osmlab/atlas/pull/723)

**Comment by reichg** [comment]
everything has been addressed! Review when ready!

File: src/main/java/org/openstreetmap/atlas/checks/validation/tag/UnknownHighwayTagCheck.java [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+        if (object instanceof Node && HighwayTag.isWayOnlyTag(object))
+        {
+            return Optional.of(this.createFlag(object,
+                    this.getLocalizedInstruction(0, object.getOsmIdentifier())));
+        }
+
+        if (object instanceof Edge && HighwayTag.isNodeOnlyTag(object))
+        {
+            return Optional.of(this.createFlag(object,
```
Comment (Bentleysb) [inline src/main/java/org/openstreetmap/atlas/checks/validation/tag/UnknownHighwayTagCheck.java:84]: Please use OsmWayWalker to gather all Edges from the same Way when creating flags.

File: src/main/java/org/openstreetmap/atlas/checks/validation/tag/UnknownHighwayTagCheck.java [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+                    this.getLocalizedInstruction(1, object.getOsmIdentifier())));
+        }
+
+        if (objectHighwayTag.isPresent()
+                && !this.isKnownHighwayTag(objectHighwayTag.get().toUpperCase(), allHighwayTags))
+        {
+            return Optional.of(this.createFlag(object,
```
Comment (Bentleysb) [inline src/main/java/org/openstreetmap/atlas/checks/validation/tag/UnknownHighwayTagCheck.java:91]: If the object is an Edge please use OsmWayWalker to gather all Edges from the same Way when creating flags.

Comment (reichg) [reply src/main/java/org/openstreetmap/atlas/checks/validation/tag/UnknownHighwayTagCheck.java:99]: addressed

File: src/main/java/org/openstreetmap/atlas/checks/validation/tag/UnknownHighwayTagCheck.java (hunk 1 above)
Comment (reichg) [reply src/main/java/org/openstreetmap/atlas/checks/validation/tag/UnknownHighwayTagCheck.java]: addressed

**Review by Bentleysb — DISMISSED** [review]

File: src/main/java/org/openstreetmap/atlas/checks/validation/tag/UnknownHighwayTagCheck.java [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+     * @param object
+     *            the atlas object supplied by the Atlas-Checks framework for evaluation
+     * @return an optional {@link CheckFlag} object that
+     */
+    @Override
+    protected Optional<CheckFlag> flag(final AtlasObject object)
+    {
+        markAsFlagged(object.getOsmIdentifier());
+        final Optional<String> objectHighwayTag = object.getTag(HighwayTag.KEY);
+        final Set<String> allHighwayTags = Arrays.stream(HighwayTag.values()).map(Enum::toString)
```
Comment (seancoulter) [inline src/main/java/org/openstreetmap/atlas/checks/validation/tag/UnknownHighwayTagCheck.java:74]: this could be a class constant

**Review by seancoulter — COMMENTED** [review]
Looks great, just noticed something that could be pulled out to avoid recalculation

File: src/main/java/org/openstreetmap/atlas/checks/validation/tag/UnknownHighwayTagCheck.java (hunk 3 above)
Comment (reichg) [reply src/main/java/org/openstreetmap/atlas/checks/validation/tag/UnknownHighwayTagCheck.java]: Should be ready when the ci completed.

**Review by Bentleysb — APPROVED** [review]

**Review by seancoulter — APPROVED** [review]

**Pull request merged**